grafana/grafana
# Elasticsearch: Add processing for raw data to backend

**What is this feature?**

Part of https://github.com/grafana/grafana/issues/54011. This PR adds processing of raw data response to backend. To do this, we have encapsulated all logic into `processDocumentResponse`.

`processDocumentResponse` processes response in the same way as it is currently done on frontend:

1. From `hits` in response, creates `docs`
2. `Docs` are flattened (multi level object are flattened to single level object where keys are connected trough `.`)
3. We create list of `propNames` (unique property names) and order them alphabetically. Timefield is exception and is added to first place.
4. Each `propName` will become dataFrame field and the values will be pushed from individual `doc`.
5. To detect dataFrame type, we look at first non-null value and set type based on it. 

The logic to create dataFrame field for each individual type seems repetitive, but I had hard time to come up with repetitive, type-safe and clean solution, so I kept it this way. But all recommendations are more than welcome!

I also added tests and snapshot test with response that includes:

- nested object that needs to be flattened
- in first doc we have field that has null value. In second that that field has actual value - making sure fata frame field has correct type. 

**Special notes for your reviewer**:

To test this
1. Run ElasticSearch data source by `make devenv sources=elastic`
2. Use `elasticsearchBackendMigration` feature toggle to toggle if query is run trough frontend/backend
3. Run `Raw data` query
4. Results run trough frontend and backend should be the same
5. Make sure you check also types of values in table
